plugins/modules/system/disk_facts.py

```diff
@@ -0,0 +1,181 @@
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: disk_facts
+version_added: "2.5.0"
+short_description: Get disk facts in the linux OS
+description:
+    - This module is to get disk facts like free space, used percentage, mount location etc in linux Ansible controller/controlled machines.
+requirements:
+  - psutil(python module)
+options:
+    name:
+        description:
+            - This is the disk name to query
+        required: false
+        default: all
+        type: str
+    filter:
+        description:
+            - This is the parameter used to filter the output
+        required: false
+        type: list
+        elements: str
+author:
+    - Saranya Sridharan(@saranyasridharan), Swetha M, Nagendra Kini, Anup Kumar, Mehmoodkha Pathan
+'''
+
+EXAMPLES = '''
+- name: Get disk facts
+  community.general.disk_facts:
+
+- name: Get Disk facts for Particular disk info
+  community.general.disk_facts:
+    name: diskname
+
+- name: Filter the output parameters for specific disk
+  community.general.disk_facts:
+    name: diskname
+    filter:
+      - mountpoint
+      - freespace
+      - totalsize
+
+- name: Filter the output parameters for all disk
+  community.general.disk_facts:
+    filter:
+      - fstype
+      - usedspace
+      - capacity_percentage
+'''
+
+
+RETURN = '''
+ansible_facts:
+    description: disk information only filtered parameter values if filter option is given.
+    returned: dictionary of disks informations
+    type: dict
+    sample: {
+        '/dev/disk1s1': {
+             'capacity_percentage': '2%'
+        }
+    }
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule
+try:
+    import psutil
+    HAS_PSUTIL = True
+except ImportError:
+    HAS_PSUTIL = False
+
+
+# To calculate the disk size
+def calculate_size(bytes, suffix="B"):
+    """
+    Scale bytes to its proper format
+    e.g:
+        1253656 => '1.20MB'
+        1253656678 => '1.17GB'
+    """
+    factor = 1024
+    for unit in ["", "K", "M", "G", "T", "P"]:
+        if bytes < factor:
+            return "{0:.2f}{1}{2}".format(bytes, unit, suffix)
+        bytes /= factor
+
+
+# To get the disk partitions information
+def disk_facts(devicename):
+    output = {}
+    partitions = psutil.disk_partitions()
+    found = False
+    for partition in partitions:
+        if partition.device == devicename:
+            found = True
+        output[partition.device] = {}
+        output[partition.device]['mountpoint'] = partition.mountpoint
+        output[partition.device]['fstype'] = partition.fstype
+        try:
+            partition_usage = psutil.disk_usage(partition.mountpoint)
```

Comment:
In case `devicename` is not `'all'`, you are calling this for **every** device, even though you only need the information for one device! This is very inefficient.

Reply:
I have used break now. is it better now?

Reply:
No, only a little bit. You are still querying information on all devices before the correct one.

Reply:
I thought of getting element index in list and then query only that when its specific disk query. But as the partiotions list has sdisk objects, I can't specify query the details without iterating.  @felixfontein

Reply:
It would help if you would decide at the beginning of the loop whether this partition is of interest, and skip it (instead of querying it's information) if it is not.